src/builder.rs

```diff
@@ -0,0 +1,295 @@
+
+use googleapis_raw::spanner::v1::{
+    result_set::{PartialResultSet, ResultSetMetadata, ResultSetStats},
+    spanner::ExecuteSqlRequest,
+    type_pb::{StructType_Field, Type, TypeCode},
+};
+
+use protobuf::{
+    well_known_types::{ListValue, NullValue, Struct, Value},
+    RepeatedField,
+};
+
+use grpcio::ClientSStreamReceiver;
+
+use crate::errors::DbError;
+
+pub type Conn = crate::connection::SpannerConnection;
+
+pub fn as_value(string_value: String) -> Value {
+    let mut value = Value::new();
+    value.set_string_value(string_value);
+    value
+}
+
+pub fn as_type(v: TypeCode) -> Type {
+    let mut t = Type::new();
+    t.set_code(v);
+    t
+}
+
+pub fn struct_type_field(name: &str, field_type: TypeCode) -> StructType_Field {
+    let mut field = StructType_Field::new();
+    field.set_name(name.to_owned());
+    field.set_field_type(as_type(field_type));
+    field
+}
+
+pub fn as_list_value(
+    string_values: impl Iterator<Item = String>,
+) -> protobuf::well_known_types::Value {
+    let mut list = ListValue::new();
+    list.set_values(RepeatedField::from_vec(
+        string_values.map(as_value).collect(),
+    ));
+    let mut value = Value::new();
+    value.set_list_value(list);
+    value
+}
+
+pub fn null_value() -> Value {
+    let mut value = Value::new();
+    value.set_null_value(NullValue::NULL_VALUE);
+    value
+}
+
+#[derive(Default)]
+pub struct ExecuteSqlRequestBuilder {
+    execute_sql: ExecuteSqlRequest,
+    params: Option<HashMap<String, Value>>,
+    param_types: Option<HashMap<String, Type>>,
+}
+
+impl ExecuteSqlRequestBuilder {
+    pub fn new(execute_sql: ExecuteSqlRequest) -> Self {
+        ExecuteSqlRequestBuilder {
+            execute_sql,
+            ..Default::default()
+        }
+    }
+
+    pub fn params(mut self, params: HashMap<String, Value>) -> Self {
+        self.params = Some(params);
+        self
+    }
+
+    pub fn param_types(mut self, param_types: HashMap<String, Type>) -> Self {
+        self.param_types = Some(param_types);
+        self
+    }
+
+    fn prepare_request(self, conn: &Conn) -> ExecuteSqlRequest {
+        let mut request = self.execute_sql;
+        request.set_session(conn.session.get_name().to_owned());
+        if let Some(params) = self.params {
+            let mut paramss = Struct::new();
+            paramss.set_fields(params);
+            request.set_params(paramss);
+        }
+        if let Some(param_types) = self.param_types {
+            request.set_param_types(param_types);
+        }
+        request
+    }
+
+    /// Execute a SQL read statement but return a non-blocking streaming result
+    pub async fn execute(self, conn: &Conn) -> Result<StreamedResultSetAsync, DbError> {
+        let stream = conn
+            .client
+            .execute_streaming_sql(&self.prepare_request(conn))?;
+        Ok(StreamedResultSetAsync::new(stream))
+    }
+
+    // Execute a DML statement, returning the exact count of modified rows
```

Comment:
Why's this commented out?

Reply:
Well, I don't recall!

Reply:
https://github.com/mozilla-services/spanner-client-rs/issues/2